pkg/util/cloud_test.go

```diff
@@ -0,0 +1,74 @@
+package util
+
+import "testing"
+
+func TestGetCloudInstanceType(t *testing.T) {
+	type args struct {
+		instanceType        string
+		instanceTypes       []string
+		defaultInstanceType string
+	}
+	tests := []struct {
+		name string
+		args args
+		want string
+	}{
+		// Add test case with instanceType="t2.small", instanceTypes=["t2.small, t2.medium"], defaultInstanceType="t2.medium"
+		{
+			name: "instanceType=t2.small, instanceTypes=[t2.small, t2.medium], defaultInstanceType=t2.medium",
+			args: args{
+				instanceType:        "t2.small",
+				instanceTypes:       []string{"t2.small", "t2.medium"},
+				defaultInstanceType: "t2.medium",
+			},
+			want: "t2.small",
+		},
+		// Add test case with instanceType="t2.small", instanceTypes=["t2.medium"], defaultInstanceType="t2.medium"
+		{
+			name: "instanceType=t2.small, instanceTypes=[t2.medium], defaultInstanceType=t2.medium",
+			args: args{
+				instanceType:        "t2.small",
+				instanceTypes:       []string{"t2.medium"},
+				defaultInstanceType: "t2.medium",
+			},
+			want: "t2.medium",
+		},
+		// Add test case with instanceType="", instanceTypes=["t2.medium"], defaultInstanceType="t2.medium"
+		{
+			name: "instanceType=, instanceTypes=[t2.medium], defaultInstanceType=t2.medium",
+			args: args{
+				instanceType:        "",
+				instanceTypes:       []string{"t2.medium"},
+				defaultInstanceType: "t2.medium",
+			},
+			want: "t2.medium",
+		},
+		// Add test case with instanceType="", instanceTypes=[], defaultInstanceType="t2.medium"
+		{
+			name: "instanceType=, instanceTypes=[], defaultInstanceType=t2.medium",
+			args: args{
+				instanceType:        "",
+				instanceTypes:       []string{},
+				defaultInstanceType: "t2.medium",
+			},
+			want: "t2.medium",
+		},
+		// Add test case with instanceType="t2.small", instanceTypes=[], defaultInstanceType="t2.medium"
```

Comment:
Thanks for adding the unit test, that helps to grok the logic. Could we conflate `instanceTypes` and `defaultInstanceType`? e.g. make the order of `instanceTypes` semantic and declare the first entry of `instanceTypes` to be the default?

If we want to have fallback logic in the future (we cannot start a VM in region X b/c of some reason) we could pick the next entry in the list. On the other hand this might be not intended (user requested spot instance, got very pricey GPU instance).

Reply:
How about handling this in a subsequent PR ? Making this change will require modifying the existing options (and any related changes).  Imho once we have an initial support for flexible instance types across the providers, we can conflate the two options into one.
But I'm open to make this change in this PR. Let me know what you prefer.

Reply:
Ah, right `InstanceType` already exists in the config, so we use that as `defaultInstanceType` if I understood it correctly. In this case we probably want to do it seperately, since it breaks existing config maps. I'll create a follow-up issue once this is merged.